Eevee.Sleep.Bot/Modules/SlashCommands/RoleManagementSlashModule.cs

```diff
@@ -0,0 +1,138 @@
+public class RoleManagementSlashModule : InteractionModuleBase<SocketInteractionContext> {
+    private async Task SendEphemeralMessageToBeDeletedAsync(string text, MessageComponent components) {
+        await Context.Interaction.RespondAsync(
+            text: text, 
+            components: components, 
+            ephemeral: true
+        );
+        
+        // Ephemeral messages cannot be deleted by the button handler, so the message containing the button should be deleted after 30 seconds.
+        await Task.Delay(TimeSpan.FromSeconds(30));
+        await Context.Interaction.DeleteOriginalResponseAsync();
+    }
+
+    [SlashCommand("display", "Select which of the role you have to display.")]
+    [UsedImplicitly]
+    public Task DisplayRoleAsync() {
+        return SendEphemeralMessageToBeDeletedAsync(
+            "Select your role to display.",
+            components: DiscordMessageMaker.MakeRoleSelectButton(
+                roles: DiscordTrackedRoleContoller.FindAllTrackedRoleIdsByRoleIds(
+                    DiscordRoleRecordContoller.FindRoleIdsByUserId(Context.User.Id)
+                ),
+                buttonId: ButtonId.RoleChanger
+            )
+        );
+    }
+
+    [SlashCommand("add", "Add the role ownership of a user.")]
+    [UsedImplicitly]
+    public Task AddRoleAsync() {
+        var user = Context.User is SocketGuildUser matchedUser ?
+            matchedUser : 
+            throw new InvalidOperationException("User is not SocketGuildUser.");
+
+        return SendEphemeralMessageToBeDeletedAsync(
+            "You already have all the roles.",
+            components: DiscordMessageMaker.MakeRoleSelectButton(
+                roles: DiscordTrackedRoleContoller.FindAllTrackedRoleIdsByRoleIds(
+                    // Find the role ids that the user does not have
+                    DiscordRoleRecordContoller
+                        .FindRoleIdsByUserId(Context.User.Id)
+                        .Except(user.Roles.Select(x => x.Id))
+                        .ToArray()
+                ),
+                buttonId: ButtonId.RoleAdder
+            )
+        );
+    }
+
+    [SlashCommand("remove", "Remove the role ownership of a user.")]
+    [UsedImplicitly]
+    public Task RemoveRoleAsync() {
+        var user = Context.User is SocketGuildUser matchedUser ?
+            matchedUser : 
+            throw new InvalidOperationException("User is not SocketGuildUser.");
+
+        return SendEphemeralMessageToBeDeletedAsync(
+            "You don't have any roles to remove.",
+            components: DiscordMessageMaker.MakeRoleSelectButton(
+                roles: DiscordTrackedRoleContoller.FindAllTrackedRoleIdsByRoleIds(
+                    // Find the role ids that the user has
+                    DiscordRoleRecordContoller
+                        .FindRoleIdsByUserId(Context.User.Id)
+                        .Intersect(user.Roles.Select(x => x.Id))
```

Comment:
Should be able to use `$in` to intersect it in the DB as well?

Reply:
REMOVE looks good.
It can base on the role the user currently has.
I think add is impossible, because it needs to know the trackedRole and the role that the user doesn't have now.

Reply:
The mongo query of `{userId: <UserID>, roleId: {$nin: <RoleIdTheUserAlreadyHas>}` should work?

Reply:
I am not comfortable creating a new function because it will not be generic
I think it would be clearer to leave the current process as it is, since it would definitely be a removeRole-only process.

Reply:
🆗 I'll mark this as resolved then